### DISCUSSION TITLE
Message doesn't send with "MarkdownV2" but works with just "Markdown"

### DISCUSSION THREAD

**Discussion by EpicLPer** [windkh/node-red-contrib-telegrambot#365] [General]
I'm using the following function which doesn't work when I use "MarkdownV2", but works just fine when I just use "Markdown".

```
msg.payload.options = { parse_mode: "MarkdownV2" };

var helpMessage = "/help - Shows this help text";
helpMessage += "\r\n\r\n";
helpMessage += "⚠️ Due to RAM constraints please do not start multiple game servers at once! Else this will most likely trigger the automatic server restart threshhold quite often! ⚠️";
helpMessage += "\r\n\r\n";
helpMessage += "*Palworld Server Commands:*";
helpMessage += "\r\n/StartPalworldServer - Starts Palworld Server";
helpMessage += "\r\n/StopPalworldServer - Stops and Disables Palworld Server";
helpMessage += "\r\n/RestartPalworldServer - Restarts Palworld Server";

helpMessage += "\r\n\r\n";
helpMessage += "*Satisfactory Server Commands:*";
helpMessage += "\r\n/StartSatisfactoryServer - Starts Satisfactory Server";
helpMessage += "\r\n/StopSatisfactoryServer - Stops and Disables Satisfactory Server";
helpMessage += "\r\n/RestartSatisfactoryServer - Restarts Satisfactory Server";

msg.payload.content = helpMessage;

return msg;
```

I'm probably missing something very minimal, my knowledge of Node-Red isn't that advanced either so yeah \^\^ Thanks for the help!

**Comment by windkh** [comment]
https://core.telegram.org/bots/api#formatting-options

@EpicLPer please have a look at the doc to understand the difference between both

**Comment by windkh** [comment]
_(Drafted by Claude, Anthropic's AI coding assistant. Posted by the maintainer.)_

@EpicLPer — a more concrete answer in case this is still useful to you (or to anyone landing here from a search): the cause is that **MarkdownV2 requires escaping a much larger set of special characters than the legacy `Markdown` mode does**. If your message contains any of these characters unescaped, Telegram rejects it with `Bad Request: can't parse entities`:

```
_  *  [  ]  (  )  ~  `  >  #  +  -  =  |  {  }  .  !
```

Anything in that list, when used as literal text (not as formatting), needs a backslash in front of it. So a message like:

```
Status update: server-1 is fine (uptime 99.9%)
```

…that works in `Markdown` mode will fail in `MarkdownV2` because of the `-`, `(`, `)`, `.`, and `%` (well, `%` is fine — but `(`, `)`, `.`, `-` aren't).

### Two ways to handle it

**Pick one parse mode and stick with it.** If you don't need MarkdownV2 features (spoilers, strikethrough, underline), plain `Markdown` is fine and doesn't need the heavy escaping. Set `msg.payload.options.parse_mode = "Markdown"`.

**Escape the characters before sending.** If you do want MarkdownV2 (for `||spoiler||`, `~~strikethrough~~`, `__underline__`), pre-process the text in a Function node before the sender:

```js
function escapeMarkdownV2(text) {
    return text.replace(/([_*\[\]()~`>#+\-=|{}.!])/g, '\\$1');
}

msg.payload.content = escapeMarkdownV2(msg.payload.content);
msg.payload.options = { parse_mode: 'MarkdownV2' };
return msg;
```

Note: this escapes **everything**, which means if you wanted `*bold*` in the input, it'd be sent as `\*bold\*` (literal asterisks, no bold). For mixed-content messages where you want some characters as formatting and others as literal, you'd escape only the literal portions and leave the formatting markup alone — typically easier to build the message piece-by-piece in code rather than try to selectively escape a single string.

**Reference**: the full rules are in [the Telegram Bot API formatting docs](https://core.telegram.org/bots/api#formatting-options) (linked by windkh on the previous reply). The character list above comes from the "MarkdownV2 style" section of that page.

If you've already solved this, treat this comment as a marker for future searchers — leaving it here so the discussion is closeable with a concrete answer rather than just a doc link.
